### ISSUE TITLE
A tale about weird syntax

### ISSUE THREAD

**Issue by siriak** [dotnet/csharplang#3430]
Recently I've found some weirdly-looking syntax and I'd like to know more about it. Unfortunately, I was unable to google it or find an explanation in books I have, so I'm writing here.
If you're eager to see why I'm here, behold:
```csharp
var obj = new ClassWithCollectionField
{
    list = { 2, 3 },
};

class ClassWithCollectionField
{
    public List<int> list = new List<int>();

    public ClassWithCollectionField()
    {
        list.Add(1);
    }
}
```
What does this code do? I did my research and I'd like to share my thoughts and findings with you.
But let me digress first.
```csharp
List<int> list; // Declaration
// The next statement uses collection initializer syntax
list = new List<int>
{
    1,
};
// And it's equivalent to
list = new List<int>(); // Initialization
list.Add(1);            // Addition of 1 to the collection
```
So far so good. Let's dissect the code of interest in a similar fashion.
```csharp
ClassWithCollectionField obj; // Declaration
// The next statement uses object initializer syntax (or maybe something else?)
obj = new ClassWithCollectionField
{
    list = { 2, 3 },
};
// And it's equivalent to
obj = new ClassWithCollectionField // Initialization
// What's next?
// = sign is used, so, based on its semantics, I've guessed that it's
obj.list = new List() { 2, 3 };
// Let's see what the next line prints out
Console.WriteLine(string.Join(' ', a.list));
// 1 2 3
```
Oops!
Do you remember we had `list.Add(1)` in ClassWithCollectionField's constructor?
So our line of interest doesn't overwrite the collection, just adds to it
To check that, let's set it to null.
Here's our updated class definition (being not initialized is equivalent to being set to null in this case):
```csharp
class ClassWithCollectionField
{
    public List<int> list;
}
```
Compiling and running, we get a NullReferenceException right here:
```csharp
obj = new ClassWithCollectionField
{
    list = { 2, 3 },
};
```
That's quite an evidence of our hypothesis as list field is the only null thing here.
So what if `list = { comma-separated stuff }` is translated into multiple `list.Add(...)`?
Unfortunately, this piece of code doesn't compile. No matter what I tried, CSharp compiler thinks that { 2, 3 } is either array initializer or a piece of garbage and in both cases, I can't write code as follows:
```csharp
List<int> list = { 2, 3 };
```

IL code shows us the same (if you're into it):
```il
    IL_001d: newobj       instance void ConsoleApp5.Program/ClassWithCollectionField::.ctor()
    IL_0022: dup
    IL_0023: ldfld        class [System.Collections]System.Collections.Generic.List`1<int32> ConsoleApp5.Program/ClassWithCollectionField::list
    IL_0028: ldc.i4.2
    IL_0029: callvirt     instance void class [System.Collections]System.Collections.Generic.List`1<int32>::Add(!0/*int32*/)
    IL_002e: nop
    IL_002f: dup
    IL_0030: ldfld        class [System.Collections]System.Collections.Generic.List`1<int32> ConsoleApp5.Program/ClassWithCollectionField::list
    IL_0035: ldc.i4.3
    IL_0036: callvirt     instance void class [System.Collections]System.Collections.Generic.List`1<int32>::Add(!0/*int32*/)
    IL_003b: nop
    IL_003c: stloc.1      // obj
```

In short:
1) What is this syntax called?
2) Why can't I use it (or can I?) with ordinary variables, only with fields/properties?
3) IMO, = sign is abused here because it doesn't mean what it used to mean in the good old days (assignment). Noone of whom I talked with regarding this code sample, expected that it'd just add 2 and 3 into the collection. What's your take on that? Extended discussions are more than welcome.

**Comment by YairHalberstadt** [comment]
Just a pointer. If you can't find out what a piece of C# code does, you can check it out on https://sharplab.io/
Also there's a reasonable explanation of what's going on here at https://stackoverflow.com/questions/16794925/nested-object-initializer-syntax

**Comment by CyrusNajmabadi** [comment]
>What is this syntax called?

Collection and array initializers.

> Why can't I use it (or can I?) with ordinary variables, only with fields/properties?

Because we built it as part of object initializers, not local initializers.

> IMO, = sign is abused here because it doesn't mean what it used to mean in the good old days (assignment)

This feature was introduced back in 2006.  So this *is* "the good old days" :)

**Comment by siriak** [comment]
@YairHalberstadt thanks for your pointer, I totally forgot about sharplab.io. I'll quote quoted text from StackOverflow for future reference.  

> 7.6.10.2 Object initializers
An object initializer consists of a sequence of member initializers, enclosed by { and } tokens and separated by commas. Each member initializer must name an accessible field or property of the object being initialized, followed by an equals sign and an expression or an object initializer or **collection initializer**. It is an error for an object initializer to include more than one member initializer for the same field or property. It is not possible for the object initializer to refer to the newly created object it is initializing.

Here are answers to some of the original questions (please correct me if I'm wrong):
1) It doesn't have a special name, it's just a collection initializer nested inside of an object initializer.
2) Ok, I get it, it's done this way because it's done this way. I can't disagree with that.

Still, my point 3 is an open question (even though this language feature is quite mature, as it turns out :) ). @CyrusNajmabadi isn't it confusing to see an equals sign that doesn't assign anything? AFAIK, this is the only place in the language where = is used in this way. (not taking into account == and => as they are different from =). Or maybe I need to view it from a different angle?

**Comment by YairHalberstadt** [comment]
> isn't it confusing to see an equals sign that doesn't assign anything? AFAIK, this is the only place in the language where = is used in this way. (not taking into account == and => as they are different from =). Or maybe I need to view it from a different angle?

I think you are right that `=` almost always means assignment (except in aliases).

However this is an old feature and changing it now would be an unbelievable breaking change :)

**Comment by 333fred** [comment]
FWIW, this confused the heck out of me the first time I saw it. But there's unfortunately not much we can do here: it's an old feature that we can't change. I'm closing this out as the question has been answered.

**Comment by CyrusNajmabadi** [comment]
> Still, my point 3 is an open question (even though this language feature is quite mature, as it turns out :) ). @CyrusNajmabadi isn't it confusing to see an equals sign that doesn't assign anything?

As confusing as seeing a `{` mean:

`a block`
`an object initializer`
`a collection initializer (including sub-initailizers)`
`an array initializer`
`a class/struct/interface body token`
`an interpolation token`
`property pattern`
`a switch body`
`an accessor list`

:)

Syntax is overloaded massively in the language.  c'est la vie.
